File: LeftRotation.cpp
//Statement
/*
A left rotation operation on an array of size  shifts each of the array's elements  unit to the left. Given an integer, , rotate the array that many steps left and return the result.
Example

d = 2
arr = [1,2,3,4,5]

After 2 rotations, arr' = [3,4,5,1,2] .

Function Description

Complete the rotateLeft function in the editor below.

rotateLeft has the following parameters:

int d: the amount to rotate by
int arr[n]: the array to rotate
Returns

int[n]: the rotated array

*/

//Solutions

//Using deque
//Time:O(n)
//Space:O(n)
#include <bits/stdc++.h>
using namespace std;

// vector<int> rotateLeft(int d, vector<int> arr) {
//     d = d%arr.size();
//     deque<int> numbers;
//     for(auto number:arr)
//         numbers.push_back(number);
//     while(d--)
//     {
//         int firstNum = numbers[0];
//         numbers.pop_front();
//         numbers.push_back(firstNum);
//     }
//     for(int x=0;x<(int)numbers.size();x++)
//         arr[x] = numbers[x];
//     return arr;
// }

//Time:O(n)
//Space:O(n)
vector<int> rotateLeft(int d, vector<int> arr) {
	d%=arr.size();
	vector<int> tempArray = arr;
    for(int x=0;x<arr.size();x++)
    {
        int number = arr[x];
        int newIndex = ((arr.size() - d) + x) % arr.size();
        tempArray[newIndex] = number;   
    }
    return tempArray;
}

//Time:O(n)
//Space:O(1)
vector<int> rotateLeft(int d,vector<int> arr)
{
    reverse(arr.begin(),arr.end());
    reverse(arr.begin()+arr.size()-d,arr.end());
    reverse(arr.begin(),arr.end()-d);
    return arr;
}

File: LeftRotation.py
# Using deque
# Time:O(n)
# Space:O(n)
from collections import deque

# def rotateLeft(d, arr):
#     d%=len(arr)
#     tempArray = deque(arr)
#     while(d):
#         first = tempArray.popleft()
#         tempArray.append(first)
#         d-=1
#     return tempArray

# Time:O(n)
# Space: O(n)
def rotateLeft(d, arr):
    d%=len(arr)
    tempArray = list(arr)
    for i in range(0,len(arr)):
        number = arr[i]
        newIndex = (len(arr)+i - d) % len(arr)
        tempArray[newIndex] = number
    return tempArray

# Time:O(n)
# Space: O(1)
def reverseParcial(i,j,arr):
    while(i < j):
        arr[i],arr[j] = arr[j],arr[i]
        i+=1
        j-=1
        
def rotateLeft(d, arr):
    d%=len(arr)
    arr.reverse()
    reverseParcial(len(arr)-d,len(arr)-1,arr)
    reverseParcial(0,len(arr)-d-1,arr)
    return arr